fvtest/jitbuildertest/IfThenElseTest.cpp

```diff
@@ -0,0 +1,103 @@
+/*******************************************************************************
+ *
+ * (c) Copyright IBM Corp. 2017, 2017
+ *
+ *  This program and the accompanying materials are made available
+ *  under the terms of the Eclipse Public License v1.0 and
+ *  Apache License v2.0 which accompanies this distribution.
+ *
+ *      The Eclipse Public License is available at
+ *      http://www.eclipse.org/legal/epl-v10.html
+ *
+ *      The Apache License v2.0 is available at
+ *      http://www.opensource.org/licenses/apache2.0.php
+ *
+ * Contributors:
+ *    Multiple authors (IBM Corp.) - initial implementation and documentation
+ ******************************************************************************/
+
+#include "JBTestUtil.hpp"
+#include "compiler/ilgen/IlBuilder.hpp"
+
+
+// Both path exist, no merge return needed
+DEFINE_BUILDER(SubIfThenElseBothPathBuilder,
+               Int64,
+               PARAM("valueA", Int64),
+               PARAM("valueB", Int64) )
+   {
+   TR::IlBuilder *elsePath = NULL;
+   TR::IlBuilder *thenPath = NULL;
+
+   IfThenElse(&thenPath, &elsePath, ConstInt64(1));
+   thenPath->Return(
+                   thenPath->Sub(
+                   thenPath->Load("valueA"),
+                   thenPath->Load("valueB")));
+   elsePath->Return(
+                   elsePath->Sub(
+                   elsePath->Load("valueB"),
+                   elsePath->Load("valueA")));
+   return true;
+   }
+
+// one path exist and will be taken, still need hardcoded Return
+DEFINE_BUILDER(SubIfThenNullBuilder,
+               Int64,
+               PARAM("valueA", Int64),
+               PARAM("valueB", Int64) )
+   {
+   TR::IlBuilder *thenPath = NULL;
+
+   IfThenElse(&thenPath, NULL, ConstInt64(1));
+   thenPath->Return(
+                   thenPath->Sub(
+                   thenPath->Load("valueA"),
+                   thenPath->Load("valueB")));
+
+   Return(Sub(
+              Load("valueA"),
+              Load("valueB")));
+   return true;
+   }
+
+// The NULL path will be taken, need hardcoded Return
+DEFINE_BUILDER(SubIfNullElseBuilder,
+               Int64,
+               PARAM("valueA", Int64),
+               PARAM("valueB", Int64) )
+   {
+   TR::IlBuilder *elsePath = NULL;
+
+   IfThenElse(NULL, &elsePath, ConstInt64(1));
+   elsePath->Return(
+                   elsePath->Sub(
+                   elsePath->Load("valueB"),
+                   elsePath->Load("valueA")));
+
+   Return(Sub(
+              Load("valueA"),
+              Load("valueB")));
+   return true;
+   }
+
+typedef int64_t (*IfLongFunctionType)(int64_t, int64_t);
+class IfThenElseTest : public JitBuilderTest {};
+
+TEST_F(IfThenElseTest, TrueBothPath)
+   {
+   IfLongFunctionType typeIfFunction;
+   ASSERT_COMPILE(TR::TypeDictionary, SubIfThenElseBothPathBuilder, typeIfFunction);
```

Comment:
I think there would be value here in calling the compiled code and making sure it returns an expected value.